File: solutions/haskell/bob/1/src/Bob.hs
module Bob (responseFor) where

import Data.Char

responseFor :: String -> String
responseFor statement
    | isQuestion withoutSpaces && isAllCaps letters = "Calm down, I know what I'm doing!"
    | isQuestion withoutSpaces = "Sure."
    | withoutSpaces == "" = "Fine. Be that way!"
    | isAllCaps letters = "Whoa, chill out!"
    | otherwise = "Whatever."

    where
        withoutSpaces = filter (not . isSpace) statement
        letters = filter isAlpha withoutSpaces
        isQuestion [] = False
        isQuestion xs = last xs == '?'
        isAllCaps [] = False
        isAllCaps [y] = isAsciiUpper y
        isAllCaps (y:ys) = isAsciiUpper y && isAllCaps ys
    

File: solutions/kotlin/bob/1/src/main/kotlin/Bob.kt
object Bob {
    const val sure = "Sure.";
    const val chill = "Whoa, chill out!";
    const val iKnow = "Calm down, I know what I'm doing!";
    const val fine = "Fine. Be that way!";
    const val whatever = "Whatever.";

    private fun isQuestion(input: String): Boolean = input.trimEnd().lastOrNull() == '?'
    private fun isScream(input: String): Boolean = input.all { !it.isLowerCase() } && input.any { it.isUpperCase() }
    private fun isBlank(input: String): Boolean = input.isBlank()

    fun hey(input: String): String {
        val screaming = isScream(input)
        val questioning = isQuestion(input)
        return when {
            isBlank(input) -> fine
            screaming && questioning -> iKnow
            screaming -> chill
            questioning -> sure
            else -> whatever
        }
    }
}
